m2cgen/interpreters/interpreter.py

```diff
@@ -4,15 +4,74 @@
 
 
 class BaseInterpreter:
+    """
+    Base class of AST interpreter. Provides single public method .interpret()
+    which takes instance of AST expression and recursively applies method
+    _do_interpret() to it.
+    """
+
+    def interpret(self, expr):
+        return self._do_interpret(expr)
+
+    # Private methods implementing Visitor pattern
+
+    def _pre_interpret_hook(self, expr, **kwargs):
+        return None, kwargs
+
+    def _do_interpret(self, expr, **kwargs):
+        # Hook which allows to override kwargs and to return custom result.
+        result, kwargs = self._pre_interpret_hook(expr, **kwargs)
+
+        # If result is empty, it means that we still need to process expr.
+        if result is not None:
+            return result
+
+        try:
+            handler = self._select_handler(expr)
+        except NotImplementedError:
+            if isinstance(expr, ast.TransparentExpr):
+                return self._do_interpret(expr.expr,  **kwargs)
+            raise
+
+        return handler(expr, **kwargs)
+
+    def _select_handler(self, expr):
+        handler_name = self._handler_name(type(expr))
+        if hasattr(self, handler_name):
+            return getattr(self, handler_name)
+        raise NotImplementedError(
+            "No handler found for {}".format(type(expr).__name__))
+
+    @staticmethod
+    def _handler_name(expr_tpe):
+        expr_name = BaseInterpreter._normalize_expr_name(expr_tpe.__name__)
+        return "interpret_" + expr_name
+
+    @staticmethod
+    def _normalize_expr_name(name):
+        return re.sub("(?!^)([A-Z]+)", r"_\1", name).lower()
+
+
+class BaseAstToCodeInterpreter(BaseInterpreter):
```

Comment:
Was `Ast` remained here on purpose? Or did you just miss this one?

Reply:
Done